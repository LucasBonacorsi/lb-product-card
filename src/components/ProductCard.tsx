import styles from "../styles/styles.module.css";
import { useProduct } from "../hooks/useProduct";
import React,{ createContext, CSSProperties, } from "react";
import {
  InitialValues,
  OnChangeArgs,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from "../interfaces/interfaces";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import ProductButtons from "./ProductButtons";

export const ProductContext = createContext({} as ProductContextProps);

export interface ProductCardProps {
  product: Product;
  children: ((args: ProductCardHandlers) => JSX.Element);
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy,
    isMaxCountReached,reset} = useProduct({
    onChange,
    product,
    value,
    initialValues
  });

  return (
    <ProductContext.Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          reset,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy
        })}
      </div>
    </ProductContext.Provider>
  );
};

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Button = ProductButtons;
