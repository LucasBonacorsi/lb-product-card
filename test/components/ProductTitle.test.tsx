import React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle, ProductCard } from '../../src/components/index';
import { product1 } from '../data/products';

describe('ProductTitle', () => {
  test('Debe mostrar el componente correctament con el titulo ', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom Product" className="Custom classname" />
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe mostrar el componente con el nmombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
