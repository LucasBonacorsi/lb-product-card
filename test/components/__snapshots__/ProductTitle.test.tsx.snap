// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle Debe mostrar el componente correctament con el titulo  1`] = `
<span
  className="productDescription Custom classname"
>
   
  Custom Product
</span>
`;

exports[`ProductTitle debe mostrar el componente con el nmombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
     
    Coffe Mug - sin imagen
  </span>
</div>
`;
